In [20]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.svm

In [29]:
import warnings
warnings.filterwarnings("ignore")

In [5]:
from sklearn.datasets import load_digits

In [13]:
from sklearn.preprocessing import StandardScaler

In [6]:
ld = load_digits()

In [8]:
df = pd.DataFrame(ld.data,columns = ld.feature_names)

In [9]:
df.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        5.0        0.0   
3        0.0        0.0        0.0        8.0  ...        9.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0        6.0       13.0       10.0        0.0   
1        0.0        0.0        0.0       11.0       16.0       10.0   
2        0.0        0.0        0.0        3.0       11.0       16.0   
3        0.0        0.0        7.0       13.0       13.0        9.0   
4        0.0        0.0        0.0        2.0       16.0        4.0   

   pixel_7_6  pixel_7_7  
0        0.0        0.0  
1        0.0        0.0  
2        9.0        0.0  
3        0.0        0.0  
4        0.0        0.0  

[5 rows x 64 columns]

In [10]:
Y = ld.target

In [12]:
Y

array([0, 1, 2, ..., 8, 9, 8])

In [14]:
scaler = StandardScaler()

In [15]:
df_scaled = scaler.fit_transform(df)

In [17]:
df_scaled

array([[ 0.        , -0.33501649, -0.04308102, ..., -1.14664746,
        -0.5056698 , -0.19600752],
       [ 0.        , -0.33501649, -1.09493684, ...,  0.54856067,
        -0.5056698 , -0.19600752],
       [ 0.        , -0.33501649, -1.09493684, ...,  1.56568555,
         1.6951369 , -0.19600752],
       ...,
       [ 0.        , -0.33501649, -0.88456568, ..., -0.12952258,
        -0.5056698 , -0.19600752],
       [ 0.        , -0.33501649, -0.67419451, ...,  0.8876023 ,
        -0.5056698 , -0.19600752],
       [ 0.        , -0.33501649,  1.00877481, ...,  0.8876023 ,
        -0.26113572, -0.19600752]])

In [24]:
from sklearn.model_selection import cross_val_score

# Cross validation score checking for logistic regression

In [30]:
cross_val_score(LogisticRegression(), df,Y, cv = 5)

array([0.92222222, 0.86944444, 0.94150418, 0.93871866, 0.89693593])

# Performing PCA

In [36]:
df.shape

(1797, 64)

In [31]:
from sklearn.decomposition import PCA

In [39]:
pca = PCA(0.95) # Using only those components which have 95% of information

In [40]:
df_pca = pca.fit_transform(df,Y)

In [41]:
df_pca.shape

(1797, 29)

# Performing GridSearchCV on SVM

In [58]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC


In [61]:
clf = GridSearchCV(SVC(), {
    'C' : [1,10,20],
    'kernel' : ['rbf', 'linear']
}, cv = 5, return_train_score=False)

clf.fit(df,Y)
clf.cv_results_

{'mean_fit_time': array([0.04301028, 0.02260385, 0.03701115, 0.02400365, 0.0398129 ,
        0.02360387]),
 'std_fit_time': array([2.53523772e-03, 7.99823680e-04, 8.91655613e-04, 2.88318948e-06,
        2.31126404e-03, 1.73899966e-03]),
 'mean_score_time': array([0.02880478, 0.00700173, 0.02560425, 0.00680161, 0.02700138,
        0.00700512]),
 'std_score_time': array([3.99721074e-04, 4.15696997e-07, 8.00852046e-04, 4.00066489e-04,
        8.99776859e-04, 6.88299201e-06]),
 'param_C': masked_array(data=[1, 1, 10, 10, 20, 20],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 1, 'kernel': 'linear'},
  {'C': 10, 'kernel': 'rbf'},
  {'C': 10, 'kernel': 'linear'},
  {'C': 20, 'ker

In [62]:
dataset = pd.DataFrame(clf.cv_results_)

In [63]:
dataset.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0       0.043010      0.002535         0.028805    3.997211e-04       1   
1       0.022604      0.000800         0.007002    4.156970e-07       1   
2       0.037011      0.000892         0.025604    8.008520e-04      10   
3       0.024004      0.000003         0.006802    4.000665e-04      10   
4       0.039813      0.002311         0.027001    8.997769e-04      20   

  param_kernel                         params  split0_test_score  \
0          rbf      {'C': 1, 'kernel': 'rbf'}           0.961111   
1       linear   {'C': 1, 'kernel': 'linear'}           0.963889   
2          rbf     {'C': 10, 'kernel': 'rbf'}           0.980556   
3       linear  {'C': 10, 'kernel': 'linear'}           0.963889   
4          rbf     {'C': 20, 'kernel': 'rbf'}           0.980556   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.944444           0.983287           0.988858           0.938719   
1           0.919444           0.966574           0.963788           0.924791   
2           0.958333           0.983287           0.988858           0.958217   
3           0.919444           0.966574           0.963788           0.924791   
4           0.958333           0.983287           0.988858           0.958217   

   mean_test_score  std_test_score  rank_test_score  
0         0.963284        0.020086                3  
1         0.947697        0.020978                4  
2         0.973850        0.012995                1  
3         0.947697        0.020978                4  
4         0.973850        0.012995                1

In [64]:
dataset[['param_C', 'param_kernel', 'mean_test_score']]

param_C param_kernel  mean_test_score
0       1          rbf         0.963284
1       1       linear         0.947697
2      10          rbf         0.973850
3      10       linear         0.947697
4      20          rbf         0.973850
5      20       linear         0.947697

In [65]:
clf.best_params_

{'C': 10, 'kernel': 'rbf'}

# Using RandomizedSearchCV

In [66]:
from sklearn.model_selection import RandomizedSearchCV

In [67]:
rs = RandomizedSearchCV(SVC(),{
    'C' : [1,5,10,20,30],
    'kernel' : ['rbf','linear']
}, cv = 5, return_train_score=False)
rs.fit(df,Y)
rs.cv_results_

{'mean_fit_time': array([0.04424996, 0.02304549, 0.03867273, 0.02342896, 0.04062557,
        0.02347112, 0.03844528, 0.02351952, 0.03799849, 0.02367535]),
 'std_fit_time': array([0.00292697, 0.00107119, 0.00129161, 0.00133878, 0.00201934,
        0.00096111, 0.00115291, 0.00100989, 0.00113036, 0.00108555]),
 'mean_score_time': array([0.02880793, 0.00685129, 0.02609382, 0.00697689, 0.02671461,
        0.00689812, 0.02604876, 0.00697684, 0.02573504, 0.00677915]),
 'std_score_time': array([6.99759292e-04, 3.78359892e-04, 6.11313617e-04, 1.83011070e-05,
        1.35298854e-03, 4.75727540e-04, 1.04778902e-03, 2.78221304e-05,
        7.33205049e-04, 3.87312325e-04]),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear',
                    'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False, False, False,
                    False, False],
        fill_value='?',
             dtype=object),
 'param_C': masked_array(

In [69]:
rs_df = pd.DataFrame(rs.cv_results_)
rs_df.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_kernel  \
0       0.044250      0.002927         0.028808        0.000700          rbf   
1       0.023045      0.001071         0.006851        0.000378       linear   
2       0.038673      0.001292         0.026094        0.000611          rbf   
3       0.023429      0.001339         0.006977        0.000018       linear   
4       0.040626      0.002019         0.026715        0.001353          rbf   

  param_C                        params  split0_test_score  split1_test_score  \
0       1     {'kernel': 'rbf', 'C': 1}           0.961111           0.944444   
1       1  {'kernel': 'linear', 'C': 1}           0.963889           0.919444   
2       5     {'kernel': 'rbf', 'C': 5}           0.983333           0.955556   
3       5  {'kernel': 'linear', 'C': 5}           0.963889           0.919444   
4      10    {'kernel': 'rbf', 'C': 10}           0.980556           0.958333   

   split2_test_score  split3_test_score  split4_test_score  mean_test_score  \
0           0.983287           0.988858           0.938719         0.963284   
1           0.966574           0.963788           0.924791         0.947697   
2           0.983287           0.988858           0.958217         0.973850   
3           0.966574           0.963788           0.924791         0.947697   
4           0.983287           0.988858           0.958217         0.973850   

   std_test_score  rank_test_score  
0        0.020086                5  
1        0.020978                6  
2        0.014024                1  
3        0.020978                6  
4        0.012995                1

In [70]:
rs_df[['param_kernel', 'param_C', 'mean_test_score']]

param_kernel param_C  mean_test_score
0          rbf       1         0.963284
1       linear       1         0.947697
2          rbf       5         0.973850
3       linear       5         0.947697
4          rbf      10         0.973850
5       linear      10         0.947697
6          rbf      20         0.973850
7       linear      20         0.947697
8          rbf      30         0.973850
9       linear      30         0.947697

In [71]:
rs.best_params_

{'kernel': 'rbf', 'C': 5}

# GridSearchCV on multiple algorithms

In [81]:
from sklearn.ensemble import RandomForestClassifier
from sklearn import linear_model

In [88]:
model_params = {
    'svm' : {
        'model' : SVC(gamma = 'auto'),
            'params' : {
                'C' : [1,5,10,20],
                'kernel' : ['rbf' , 'linear']
            }
    },
    'random_forest' : {
        'model' : RandomForestClassifier(),
        'params' : {
            'n_estimators' : [20,30,40,50]
        }
    },
    'log_reg' : { 
        'model' : linear_model.LogisticRegression(),
                 'params' : {
                     'C' : [1,5,10,20]
                 }
    }
}
    

In [89]:
model_params.items()

dict_items([('svm', {'model': SVC(gamma='auto'), 'params': {'C': [1, 5, 10, 20], 'kernel': ['rbf', 'linear']}}), ('random_forest', {'model': RandomForestClassifier(), 'params': {'n_estimators': [20, 30, 40, 50]}}), ('log_reg', {'model': LogisticRegression(), 'params': {'C': [1, 5, 10, 20]}})])

In [84]:
scores = []

for model_name, mp in model_params.items():
    clf =  GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf.fit(ld.data, ld.target)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })
    
df = pd.DataFrame(scores,columns=['model','best_score','best_params'])
df

model  best_score                   best_params
0            svm    0.947697  {'C': 1, 'kernel': 'linear'}
1  random_forest    0.936026          {'n_estimators': 50}
2        log_reg    0.913765                      {'C': 1}

# Practice

In [90]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier

In [112]:
model_2 = {'Gaussian' : {
                'model' : GaussianNB(),
                   'params' : 
                   {}
                   },
        'Multinomial' : { 
                'model' :MultinomialNB(),
                  'params' : {}
                  },
        'DT' : { 
                'model' : DecisionTreeClassifier(),
                  'params' : {
                      'criterion' : ['gini', 'entropy']
                  }
              }
         }

In [114]:
score= []
for model_name , mp in model_2.items():
    gs = GridSearchCV(mp['model'], mp['params'], cv = 5, return_train_score=False)
    gs.fit(ld.data,ld.target)
    score.append({
        'model_name' : model_name,
        'best_params' : clf.best_params_,
        'score' : clf.best_score_,
    })

In [115]:
score

[{'model_name': 'Gaussian',
  'best_params': {'C': 1},
  'score': 0.9137650882079852},
 {'model_name': 'Multinomial',
  'best_params': {'C': 1},
  'score': 0.9137650882079852},
 {'model_name': 'DT', 'best_params': {'C': 1}, 'score': 0.9137650882079852}]

In [116]:
df = pd.DataFrame(score,)

In [117]:
df.head()

model_name best_params     score
0     Gaussian    {'C': 1}  0.913765
1  Multinomial    {'C': 1}  0.913765
2           DT    {'C': 1}  0.913765

## The conclusion is [svm with accuracy of 0.947697	is performing the best among all algorithms with these parameters {'C': 1, 'kernel': 'linear'} ] 